### PR TITLE
add compare query version support

### PR DESCRIPTION
**Pull request by spasovski**
How to trigger:

The modal itself:

### CODE REVIEW

File: redash/handlers/queries.py [hunk 1]
```diff
@@ -264,3 +265,23 @@ def post(self, query_id):
...
+            models.Change.object_type == 'queries',
+            models.Change.object_id == query_id)
+        page = request.args.get('page', 1, type=int)
+        page_size = request.args.get('page_size', 25, type=int)
+        return paginate(results, page, page_size, lambda c: c.to_dict(full=False))
+
+
+class QueryVersionResource(BaseResource):
+    @require_permission('view_query')
+    def get(self, query_id, version_id):
```
Comment (arikfr) [inline redash/handlers/queries.py]: Why not use the `Change` object's primary key? It will make the query much more efficient and simpler.

File: redash/handlers/queries.py [hunk 2]
```diff
@@ -264,3 +265,23 @@ def post(self, query_id):
         parameter_values = collect_parameters_from_request(request.args)
 
         return run_query(query.data_source, parameter_values, query.query_text, query.id)
+
+
+class QueryVersionListResource(BaseResource):
+    @require_permission('view_query')
+    def get(self, query_id):
+        results = models.Change.query.filter(
```
Comment (arikfr) [inline redash/handlers/queries.py]: This logic (finding changes for an object) should be in the model (i.e. a method like `Change.for` that takes the model).

Also you need to fetch the query first & check for proper permissions. You can see [an example in alerts](https://github.com/getredash/redash/blob/master/redash/handlers/alerts.py#L47-L49).

File: client/app/pages/queries/compare-query-dialog.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+          span.appendChild(document.createTextNode(part.value));
+          fragment.appendChild(span);
+        });
+
+        return fragment;
+      }
+
+      document.getElementById('current-query-diff').appendChild(getDiffContent(currentDiff));
+      document.getElementById('previous-query-diff').appendChild(getDiffContent(previousDiff));
+    });
```
Comment (arikfr) [inline client/app/pages/queries/compare-query-dialog.js]: This whole logic belongs in the template. You can easily express it with Angular:

```html
<span ng-class="{'diff-added': part.added, 'diff-removed': part.removed}" ng-repeat="part in diff">
{{part.value}}
</span>
```

File: client/app/assets/css/redash.css [hunk 4]
```diff
@@ -667,3 +667,36 @@ div.table-name:hover {
 .m-2{
   margin:2px;
 }
+
+/* Compare Query Version container */
+/* Offers slight visual improvement (alignment) to modern UAs */
```
Comment (arikfr) [inline client/app/assets/css/redash.css]: You can place this along side the component. [Example](https://github.com/getredash/redash/blob/master/client/app/components/app-header/index.js#L5).

File: client/app/pages/queries/query.html [hunk 5]
```diff
@@ -74,6 +74,22 @@
...
+        <select id="version-choice">
+          <option ng-repeat="version in versions" value="{{version.object_version}}">
+            {{version.object_version}}
+          </option>
+        </select>
+        <button class="btn btn-s btn-default" ng-click="compareQueryVersion()">
+          <span class="zmdi zmdi-play"></span>
+          Compare
+        </button>
+      </div>
```
Comment (arikfr) [inline client/app/pages/queries/query.html]: Next time, it will be great if you could include a screenshot of all the UI changes. But based on the markup -- I think it will take too much space.

How about linking the `updated at` timestamp to a UI similar to what Dropbox does in their Paper product:

image

Basically they show a list of all the recent changes. It's both efficient on the screen real estate, but also efficient in reviewing the changes, as you don't have to pick each individual version.

You just need to change the changes list API to return the full change set and then we save on multiple API calls. Another thing to consider is that not all changes are changes to the query text, so need to filter those out (can be done in the client side).

Comment (spasovski) [inline client/app/pages/queries/query.html]: Yea good call on more screenshots I was a bit too focused on the modal. The dropbox example wasn't the way we'd intended to do it though I can see some of its merits - we'd gone for a more general "compare current to..." approach. 

This change would require a more significant time investment so we might need to see if it fits Mozilla's needs. Tagging @rafrombrc on this.

File: client/app/pages/queries/compare-query-dialog.js (hunk 3 above)
Comment (spasovski) [inline client/app/pages/queries/compare-query-dialog.js]: Good call it shaved quite a few lines of JS.

File: client/app/pages/queries/query.html (hunk 5 above)
Comment (rafrombrc) [inline client/app/pages/queries/query.html]: Hrm. I do like the UI that you're proposing, @arikfr. There are some cases where it might be a bit unwieldy.. if a query is >100 lines of code and has multiple significant edits then the single page stream might be a bit long, but even so we can probably work around that by having the diffs be collapsible or something.

Might you be open to proceeding with our current UI and then treating this change as a separate issue?

Comment (arikfr) [inline client/app/pages/queries/query.html]: Regardless of how we implement the preview in the modal, the triggering UI can't be as it is in this pull request. Maybe a middle ground is to have the "updated at" timestamp trigger the comparison UI, and in the modal window show the version picker (instead of having it on the main page). This way the impact is minimal.

**Comment by spasovski** [comment]
Thanks for the comments - updated for re-review.

File: client/app/pages/queries/view.js [hunk 6]
```diff
@@ -254,14 +254,28 @@ function QueryViewCtrl($scope, Events, $route, $routeParams, $location, $window,
 
     updateSchema();
     $scope.dataSource = find($scope.dataSources, ds => ds.id === $scope.query.data_source_id);
-    $scope.executeQuery();
```
Comment (alison985) [inline client/app/pages/queries/view.js]: Are you sure removing this line doesn't break anything?

File: package.json [hunk 7]
```diff
@@ -38,6 +38,7 @@
     "d3": "^3.5.17",
     "d3-cloud": "^1.2.1",
     "debug": "^2.2.0",
+    "diff": "^3.2.0",
```
Comment (alison985) [inline package.json]: Given the issue we were having, should this be 3.3.0?

File: redash/handlers/queries.py [hunk 8]
```diff
@@ -271,3 +273,19 @@ def post(self, query_id):
...
+
+
+class QueryVersionListResource(BaseResource):
+    @require_permission('view_query')
+    def get(self, query_id):
+        results = models.Change.list_versions(models.Query.get_by_id(query_id))
+        page = request.args.get('page', 1, type=int)
+        page_size = request.args.get('page_size', 25, type=int)
+
+        return paginate(results, page, page_size, lambda q: q.to_dict(full=False))
```
Comment (alison985) [inline redash/handlers/queries.py]: I don't see paginate options in the UI/html so I'm not sure how this paginate will react at query version 26?

File: redash/models.py [hunk 9]
```diff
@@ -166,10 +166,6 @@ class ChangeTrackingMixin(object):
     skipped_fields = ('id', 'created_at', 'updated_at', 'version')
     _clean_values = None
 
-    def __init__(self, *a, **kw):
```
Comment (alison985) [inline redash/models.py]: Why remove this? How is it related?

File: redash/models.py [hunk 10]
```diff
@@ -194,13 +190,19 @@ def record_changes(self, changed_by):
...
+                prev = self._clean_values[col.name]
+                current = getattr(self, attr.key)
+                if prev != current:
+                    changes[col.name] = {'previous': prev, 'current': current}
 
-        db.session.add(Change(object=self,
-                              object_version=self.version,
-                              user=changed_by,
-                              change=changes))
+        if changes:
```
Comment (alison985) [inline redash/models.py]: What do you return if this is false?

File: redash/models.py [hunk 11]
```diff
@@ -947,6 +949,7 @@ def fork(self, user):
         kwargs = {a: getattr(self, a) for a in forked_list}
         forked_query = Query.create(name=u'Copy of (#{}) {}'.format(self.id, self.name),
                                     user=user, **kwargs)
+        forked_query.record_changes(changed_by=user)
```
Comment (alison985) [inline redash/models.py]: This is the only place you refer to the concept of forking. Curious: what ends up getting written to the `changes` database table in this case?

File: tests/handlers/test_queries.py [hunk 12]
```diff
@@ -158,3 +158,30 @@ def test_refresh_query_you_dont_have_access_to(self):
         user = self.factory.create_user(group_ids=[group.id])
         response = self.make_request('post', self.path, user=user)
         self.assertEqual(403, response.status_code)
+
```
Comment (alison985) [inline tests/handlers/test_queries.py]: You could also add tests for QueryVersionListResource, for the fork of a query, and for a revert if you wanted.

File: redash/handlers/dashboards.py [hunk 13]
```diff
@@ -57,6 +57,7 @@ def post(self):
                                      user=self.current_user,
                                      is_draft=True,
                                      layout='[]')
+        dashboard.record_changes(changed_by=self.current_user)
```
Comment (alison985) [inline redash/handlers/dashboards.py]: I didn't see any additional dashboard references. Is recording a change to a dashboard already supported?

File: redash/models.py [hunk 14]
```diff
@@ -748,7 +750,7 @@ def should_schedule_next(previous_iteration, now, schedule, failures):
 
 class Query(ChangeTrackingMixin, TimestampMixin, BelongsToOrgMixin, db.Model):
     id = Column(db.Integer, primary_key=True)
-    version = Column(db.Integer, default=1)
+    version = Column(db.Integer, default=0)
```
Comment (alison985) [inline redash/models.py]: Are there any current instances that have this set to 1 that would get an upgrade bug/unexpected behavior because of this?

File: redash/models.py [hunk 15]
```diff
@@ -1075,7 +1078,6 @@ def to_dict(self, full=True):
             'id': self.id,
             'object_id': self.object_id,
             'object_type': self.object_type,
-            'change_type': self.change_type,
```
Comment (alison985) [inline redash/models.py]: Why is this removed?

File: client/app/pages/queries/view.js (hunk 6 above)
Comment (alison985) [inline client/app/pages/queries/view.js]: Tagging @ashort to answer my python-related questions for this PR.

File: redash/handlers/dashboards.py (hunk 13 above)
Comment (washort) [inline redash/handlers/dashboards.py]: Yes, `Dashboard` subclasses `ChangeTrackingMixin`.

File: redash/models.py (hunk 10 above)
Comment (washort) [inline redash/models.py]: `None`

File: redash/models.py (hunk 14 above)
Comment (washort) [inline redash/models.py]: No, because defaults only affect new instances, and the bugfixes in `ChangeTrackingMixin` result in the version number being bumped immediately.

File: redash/models.py (hunk 11 above)
Comment (washort) [inline redash/models.py]: The initial change, same as any other creation of a query or dashboard

File: redash/models.py (hunk 15 above)
Comment (washort) [inline redash/models.py]: There's no attribute named `change_type`.

File: redash/handlers/queries.py (hunk 8 above)
Comment (spasovski) [inline redash/handlers/queries.py]: Yep really good catch I think an earlier draft was going to have a paginated versions list.

**Comment by arikfr** [comment]
As there is a new pull request for the same feature from @spasovski, closing this one.

**Comment by shakismyname** [comment]
@arikfr Hi there! Just quickly curious - I saw that the pr got merged into the [mozilla fork](https://github.com/mozilla/redash/pull/149). But it appears that #25 is still open. Sounds like you're all still designing/working on it?

Thanks!
